src/threadNames.cpp

```diff
@@ -0,0 +1,48 @@
+#include "threadNames.h"
+#include "profiler.h"
+
+ThreadNames ThreadNames::_instance;
+
+void ThreadNames::update(jthread thread) {
+    MutexLocker ml(_lock);
+    if (!VMThread::available()) {
+        return;
+    }
+    JNIEnv *env = VM::jni();
+    if (_eetop == NULL) {
+        _threadClass = env->FindClass("java/lang/Thread");
+        if (_threadClass == NULL) {
+            return;
+        }
+        _eetop = env->GetFieldID(_threadClass, "eetop", "J");
+        if (_eetop == NULL) {
+            return;
+        }
+    }
+    VMThread *vm_thread = (VMThread *)(uintptr_t)env->GetLongField(thread, _eetop);
+    jvmtiThreadInfo info;
+    jvmtiEnv *jvmti = VM::jvmti();
+    if (jvmti->GetThreadInfo(thread, &info) != 0 || info.name == NULL) {
+        return;
+    }
+    _storage[vm_thread->osThreadId()] = std::string(info.name);
+    jvmti->Deallocate((unsigned char *)info.name);
+}
+
+NamesMap ThreadNames::getNames() {
+    MutexLocker ml(_lock);
+    std::map<int, std::string> result = _storage;
+    return result;
```

Comment:
Is it really required to copy the whole map? We need to update the map only while profiler is running, but read it only after profiler is stopped.

Reply:
It is still possible that thread related events will be called after profiler is stopped, so I think it is better to copy the whole map and be safe.